### PR TITLE
[BEAM-2982] Use the SubscriptionProvider in PubsubUnboundedSource

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
During expansion, a ValueProvider may not be accessible. This ensures
that if the subscription is based on a value provider, it will only be
evaluated when that ValueProvider is bound, rather than at construction
time.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @bjchambers

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 1]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
     assertNotNull(DisplayData.from(read));
   }
 
+  /**
+   * {@link PipelineOptions} with a ValueProvider.
+   */
+  public interface RuntimeSubscriptionOptions extends PipelineOptions {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Can you use TestPipeline.newProvider() instead?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 2]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
...
+
+  @Test
+  public void testValueProviderSubscription() {
+    RuntimeSubscriptionOptions options =
+        PipelineOptionsFactory.create().as(RuntimeSubscriptionOptions.class);
+    options.setValue(StaticValueProvider.of("projects/project/subscriptions/subscription"));
+    ValueProvider<String> subscription = options.getValue();
+    Read<String> pubsubRead = PubsubIO.readStrings().fromSubscription(subscription);
+    Pipeline.create().apply(pubsubRead);
+    assertThat(pubsubRead.getSubscriptionProvider(), not(nullValue()));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Lines 168-172 seem unnecessary - what could we possibly get wrong to make those lines fail?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 3]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
...
+    void setValue(ValueProvider<String> provider);
+  }
+
+  @Test
+  public void testValueProviderSubscription() {
+    RuntimeSubscriptionOptions options =
+        PipelineOptionsFactory.create().as(RuntimeSubscriptionOptions.class);
+    options.setValue(StaticValueProvider.of("projects/project/subscriptions/subscription"));
+    ValueProvider<String> subscription = options.getValue();
+    Read<String> pubsubRead = PubsubIO.readStrings().fromSubscription(subscription);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: You could simply pass a StaticValueProvider here

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 4]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
...
+        pubsubRead.getSubscriptionProvider().get().asPath(),
+        equalTo(options.getValue().get()));
+  }
+
+  @Test
+  public void testRuntimeValueProviderSubscription() {
+    RuntimeSubscriptionOptions options =
+        PipelineOptionsFactory.create().as(RuntimeSubscriptionOptions.class);
+    ValueProvider<String> subscription = options.getValue();
+    Read<String> pubsubRead = PubsubIO.readStrings().fromSubscription(subscription);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: And pass pipeline.newProvider("...") here

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 5]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
...
+    Pipeline.create().apply(pubsubRead);
+    assertThat(pubsubRead.getSubscriptionProvider(), not(nullValue()));
+    assertThat(pubsubRead.getSubscriptionProvider().isAccessible(), is(false));
+  }
+
+  @Test
+  public void testValueProviderTopic() {
+    RuntimeSubscriptionOptions options =
+        PipelineOptionsFactory.create().as(RuntimeSubscriptionOptions.class);
+    options.setValue(StaticValueProvider.of("projects/project/topics/topic"));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Same for these two tests

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java [hunk 6]
```diff
@@ -140,6 +149,66 @@ public void testNullSubscription() {
...
+  public void testRuntimeValueProviderTopic() {
+    RuntimeSubscriptionOptions options =
+        PipelineOptionsFactory.create().as(RuntimeSubscriptionOptions.class);
+    ValueProvider<String> topic = options.getValue();
+    Read<String> pubsubRead = PubsubIO.readStrings().fromTopic(topic);
+    Pipeline.create().apply(pubsubRead);
+    assertThat(pubsubRead.getTopicProvider(), not(nullValue()));
+    assertThat(pubsubRead.getTopicProvider().isAccessible(), is(false));
+  }
+
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Seems like we still don't have runnable pipeline tests for PubsubIO?..

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 5 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 4 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 3 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 1 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: Yes. Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 2 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: I figure it's a reasonable sanity check; there is some level of indirection in translating the actual subscription string to the "SubscriptionPath"

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java (hunk 6 above)
Comment (tgroh) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIOTest.java]: That is my understanding yes.

**Comment by bjchambers** [comment]
LGTM

**Pull request closed without merging**